File: temporal-Java/SendSignal.java
// ── File: .../temporal/SendSignal.java
//
// PURPOSE: Sends an approve or cancel signal to a
//          running workflow. Run in a separate terminal.

package com.example.temporal;

import com.example.temporal.workflows.OrderWorkflow;
import io.temporal.client.WorkflowClient;
import io.temporal.serviceclient.WorkflowServiceStubs;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class SendSignal {

    private static final Logger log =
        LoggerFactory.getLogger(SendSignal.class);

    public static void main(String[] args) {

        // Usage: pass "approve" or "cancel" as first arg
        // Default: approve
        String action = (args.length > 0) ? args[0] : "approve";

        WorkflowServiceStubs svc =
            WorkflowServiceStubs.newLocalServiceStubs();
        WorkflowClient client = WorkflowClient.newInstance(svc);

        // Get a stub to the already-running workflow
        OrderWorkflow workflow =
            client.newWorkflowStub(OrderWorkflow.class, "order-001");

        // Query current status first
        String status = workflow.getStatus();
        log.info("Current workflow status: {}", status);

        // Send signal
        if ("cancel".equalsIgnoreCase(action)) {
            workflow.cancelOrder("Customer requested cancellation");
            log.info("Sent CANCEL signal to order-001");
        } else {
            workflow.approveOrder();
            log.info("Sent APPROVE signal to order-001");
        }
    }
}

File: temporal-project/client-server/src/send_signal.py
"""
client-server/send_signal.py

PURPOSE:
  Send a Signal to a running workflow, then
  Query its current status.

  Runs on the Client server.
  Uses ONLY shared/ imports — no worker-side code.

  Signals and Queries both go through the Temporal
  Server. The Client never contacts the Worker directly.

USAGE:
  python send_signal.py order-001           # approve
  python send_signal.py order-001 cancel    # cancel
  python send_signal.py order-001 status    # query only

FLOW:
  Client ──signal/query──▶ Temporal Server
                                │
                           routes to Worker
                                │
                           Worker executes
                           signal/query handler
                                │
                           result returned
                           to Client
"""

import asyncio
import os
import sys
from dotenv import load_dotenv
from temporalio.client import Client

from order_shared.workflow_iface import OrderWorkflow

load_dotenv()


async def main(workflow_id: str, action: str) -> None:
    temporal_host = os.environ.get(
        "TEMPORAL_HOST", "localhost:7233"
    )
    namespace = os.environ.get(
        "TEMPORAL_NAMESPACE", "default"
    )

    client = await Client.connect(
        temporal_host,
        namespace=namespace,
    )

    # Get a handle to the running workflow by its ID.
    # This does NOT start a new workflow.
    handle = client.get_workflow_handle(workflow_id)

    if action == "cancel":
        print(f"Sending CANCEL signal to {workflow_id}")
        await handle.signal(OrderWorkflow.cancel)

    elif action == "status":
        # Query only — no signal sent
        pass

    else:
        # Default action: approve
        print(f"Sending APPROVE signal to {workflow_id}")
        await handle.signal(OrderWorkflow.approve)

    # Query current status after the action
    status = await handle.query(OrderWorkflow.get_status)

    print()
    print(f"Workflow ID : {workflow_id}")
    print(f"Status      : {status.status}")
    print(f"Approved    : {status.approved}")
    print(f"Cancelled   : {status.cancelled}")
    print()
    print("Check UI: http://localhost:8080")


if __name__ == "__main__":
    wf_id = sys.argv[1] if len(sys.argv) > 1 else "order-001"
    act = sys.argv[2] if len(sys.argv) > 2 else "approve"
    asyncio.run(main(wf_id, act))